### ISSUE TITLE
Performance regression in _query_hf_dataset when using delta_timestamps (~40x slower)

### ISSUE THREAD

**Issue by jianzhou0420** [huggingface/lerobot#2895]
**Title:** Performance regression in _query_hf_dataset when using delta_timestamps (~40x slower)

---

## Summary

LeRobot v3.0's `_query_hf_dataset` implementation has a severe performance regression when querying multiple indices (e.g., for action horizons with `delta_timestamps`). The current implementation is **~40x slower** than using `select()`.

## Root Cause

In commit `a4aa3164` ("fix data access bottleneck"), the implementation changed from using `select()` to direct indexing:

```python
# Old approach (fast) - used select()
torch.stack(self.hf_dataset.select(q_idx)[key])

# New approach (slow) - direct indexing
# In _query_hf_dataset, it tries column-first access which loads ENTIRE column
torch.stack(self.hf_dataset[key][indices])  # This loads entire column first (~3s)
# Then falls back to row-first
torch.stack(self.hf_dataset[indices][key])  # Still slow compared to select()
```

The column-first approach `hf_dataset[key][indices]` loads the entire column into memory before indexing, which is extremely slow. Even after falling back to row-first access, direct indexing `hf_dataset[indices][key]` is still ~40x slower than using `select()`.

## Benchmark Results

Running on a parquet-based dataset with `delta_timestamps` for 50-frame action horizon:

```
Benchmarking dataset: JianZhou0420/libero_openvla_LeRobotv3_0
  Num samples: 20
  Action horizon: 50

  FPS: 10
  Total frames: 273465

============================================================
BENCHMARK RESULTS
============================================================
1. hf_dataset[i] (single frame):
   0.152s total, 7.6ms per sample

2. hf_dataset.select(indices)[key] (OpenPi's fast approach):
   0.128s total, 6.4ms per sample

3. hf_dataset[indices][key] (LeRobot v3.0's slow approach):
   5.253s total, 262.6ms per sample

4. LeRobotDataset[i] with delta_timestamps (what training uses):
   6.980s total, 349.0ms per sample

============================================================
ANALYSIS
============================================================
select() is 40.6x faster than direct indexing

Estimated time per epoch with current LeRobotDataset:
  26.5 hours (1591 minutes)

Estimated time per epoch with select() fix:
  0.56 hours (34 minutes)
```

| Method | Time per sample | Relative |
|--------|----------------|----------|
| `hf_dataset.select(indices)[key]` | 6.4ms | 1x (baseline) |
| `hf_dataset[indices][key]` | 262.6ms | 41x slower |
| `LeRobotDataset[i]` with delta_timestamps | 349.0ms | 55x slower |

**Impact:** Training time increases from ~34 minutes to ~26.5 hours per epoch.

## Use Case

This affects Pi0/OpenPi style action chunking where we need to query 50-frame action horizons using `delta_timestamps`. This is a common pattern for diffusion-based action prediction models.

```python
# Pi0/OpenPi action chunking pattern
delta_timestamps = {"actions": [t / fps for t in range(50)]}  # 50-frame horizon
dataset = LeRobotDataset(repo_id, delta_timestamps=delta_timestamps)

# Each __getitem__ call queries 50 indices - very slow with current implementation
sample = dataset[i]
```

## Suggested Fix

Use `select()` for multi-index queries in `_query_hf_dataset`:

```python
# Fast approach
torch.stack(self.hf_dataset.select(indices)[key])
```

## Benchmark Script

<details>
<summary>Click to expand benchmark_dataset_loading.py</summary>

```python
"""Benchmark LeRobotDataset loading speed.

Compares different data access methods to measure performance impact
of LeRobot v3.0's slow _query_hf_dataset implementation.

Usage:
    python scripts/benchmark_dataset_loading.py
    python scripts/benchmark_dataset_loading.py --num-samples 100
    python scripts/benchmark_dataset_loading.py --repo-id your/dataset
"""

import argparse
import time

import torch


def benchmark_hf_dataset_direct(dataset, num_samples: int) -> float:
    """Benchmark direct hf_dataset access (no delta_timestamps processing)."""
    start = time.time()
    for i in range(num_samples):
        _ = dataset.hf_dataset[i]
    return time.time() - start


def benchmark_hf_dataset_select(dataset, num_samples: int, horizon: int = 50) -> float:
    """Benchmark hf_dataset.select() for action horizon queries."""
    start = time.time()
    for i in range(num_samples):
        indices = list(range(i, min(i + horizon, len(dataset))))
        _ = torch.stack(dataset.hf_dataset.select(indices)['actions'])
    return time.time() - start


def benchmark_hf_dataset_direct_indexing(dataset, num_samples: int, horizon: int = 50) -> float:
    """Benchmark hf_dataset[indices][key] (LeRobot v3.0's slow approach)."""
    start = time.time()
    for i in range(num_samples):
        indices = list(range(i, min(i + horizon, len(dataset))))
        _ = torch.stack(dataset.hf_dataset[indices]['actions'])
    return time.time() - start


def benchmark_lerobot_dataset(dataset_with_delta, num_samples: int) -> float:
    """Benchmark LeRobotDataset with delta_timestamps (full __getitem__)."""
    start = time.time()
    for i in range(num_samples):
        _ = dataset_with_delta[i]
    return time.time() - start


def main():
    parser = argparse.ArgumentParser(description="Benchmark LeRobotDataset loading speed")
    parser.add_argument("--repo-id", default="JianZhou0420/libero_openvla_LeRobotv3_0",
                        help="HuggingFace dataset repo ID")
    parser.add_argument("--num-samples", type=int, default=50,
                        help="Number of samples to benchmark")
    parser.add_argument("--horizon", type=int, default=50,
                        help="Action horizon for delta_timestamps")
    args = parser.parse_args()

    print(f"Benchmarking dataset: {args.repo_id}")
    print(f"  Num samples: {args.num_samples}")
    print(f"  Action horizon: {args.horizon}")
    print()

    # Import here to see any warnings
    from lerobot.datasets.lerobot_dataset import LeRobotDataset, LeRobotDatasetMetadata

    # Load metadata
    meta = LeRobotDatasetMetadata(args.repo_id)
    fps = meta.fps
    print(f"  FPS: {fps}")
    print(f"  Total frames: {meta.total_frames}")
    print()

    # Load dataset WITHOUT delta_timestamps
    print("Loading dataset without delta_timestamps...")
    dataset_simple = LeRobotDataset(args.repo_id)

    # Load dataset WITH delta_timestamps
    print("Loading dataset with delta_timestamps...")
    delta_timestamps = {"actions": [t / fps for t in range(args.horizon)]}
    dataset_with_delta = LeRobotDataset(args.repo_id, delta_timestamps=delta_timestamps)

    print()
    print("=" * 60)
    print("BENCHMARK RESULTS")
    print("=" * 60)

    # Benchmark 1: Direct hf_dataset access
    elapsed = benchmark_hf_dataset_direct(dataset_simple, args.num_samples)
    print(f"1. hf_dataset[i] (single frame):")
    print(f"   {elapsed:.3f}s total, {elapsed/args.num_samples*1000:.1f}ms per sample")
    print()

    # Benchmark 2: select() method (fast)
    elapsed = benchmark_hf_dataset_select(dataset_simple, args.num_samples, args.horizon)
    print(f"2. hf_dataset.select(indices)[key] (OpenPi's fast approach):")
    print(f"   {elapsed:.3f}s total, {elapsed/args.num_samples*1000:.1f}ms per sample")
    print()

    # Benchmark 3: Direct indexing (slow - LeRobot v3.0's approach)
    elapsed = benchmark_hf_dataset_direct_indexing(dataset_simple, args.num_samples, args.horizon)
    print(f"3. hf_dataset[indices][key] (LeRobot v3.0's slow approach):")
    print(f"   {elapsed:.3f}s total, {elapsed/args.num_samples*1000:.1f}ms per sample")
    print()

    # Benchmark 4: Full LeRobotDataset with delta_timestamps
    elapsed = benchmark_lerobot_dataset(dataset_with_delta, args.num_samples)
    print(f"4. LeRobotDataset[i] with delta_timestamps (what training uses):")
    print(f"   {elapsed:.3f}s total, {elapsed/args.num_samples*1000:.1f}ms per sample")
    print()

    print("=" * 60)
    print("ANALYSIS")
    print("=" * 60)

    # Calculate speedup
    t_select = benchmark_hf_dataset_select(dataset_simple, 10, args.horizon) / 10
    t_direct = benchmark_hf_dataset_direct_indexing(dataset_simple, 10, args.horizon) / 10
    speedup = t_direct / t_select if t_select > 0 else 0

    print(f"select() is {speedup:.1f}x faster than direct indexing")
    print()

    # Estimate full training impact
    total_frames = meta.total_frames
    t_per_sample = elapsed / args.num_samples
    estimated_epoch_time = total_frames * t_per_sample
    print(f"Estimated time per epoch with current LeRobotDataset:")
    print(f"  {estimated_epoch_time/3600:.1f} hours ({estimated_epoch_time/60:.0f} minutes)")
    print()

    # Estimate with fix
    t_fast = benchmark_hf_dataset_select(dataset_simple, args.num_samples, args.horizon) / args.num_samples
    estimated_fast_epoch = total_frames * t_fast
    print(f"Estimated time per epoch with select() fix:")
    print(f"  {estimated_fast_epoch/3600:.2f} hours ({estimated_fast_epoch/60:.0f} minutes)")


if __name__ == "__main__":
    main()
```

</details>

## Environment

- LeRobot version: v3.0 (latest main)
- Dataset format: parquet (images stored in parquet, not video)
- Python: 3.10
- Platform: Linux

**Comment by Qu3tzal** [comment]
While this is a real performance regression I don't think it has any impact on training time (the only time you would use the dataset), since the loading time is still smaller than the training time and the prefetching mechanism will make them concurrent?

**Comment by imstevenpmwork** [comment]
Hello @jianzhou0420 , thanks for taking a look !

Quick question: Isn't this only a problem when we have a massive amount of rows (or very high-res images)? Usually we have a table that stretches horizontally as we have just a few dataset features and a lot of samples.

Anyways I'm tempted to apply the changes as this was the case before the migration to V3 as explained in: https://github.com/huggingface/lerobot/pull/2408

**Comment by lhoestq** [comment]
hi, please update `datasets` and you should be fine

in old `datasets` version `ds[column]` would load the full column in memory but this is not longer the case

**Comment by Rick0525** [comment]
Confirming this regression, and adding a root-cause analysis from reading the `datasets` source. It also explains why, unfortunately, neither of the two suggestions above (reverting to `select()`, or updating `datasets`) fixes it for image-type datasets.

**TL;DR**: every delta query ends up fetching entire rows, decoding every embedded camera PNG just to read a small float column. This happens because LeRobot always sets a custom transform (`hf_dataset.set_transform(hf_transform_to_torch)` in `dataset_reader.py`), and since `datasets` 4.4 the lazy-`Column` fast path is disabled for datasets with a custom format. Building an explicit per-key `select_columns` view at the LeRobot level restores true column access: **~200×** faster on the query itself, **46×** end-to-end per sample in our training runs.

## Why both current branches are equally slow

In `datasets` ≥ 4.4, `hf_dataset[key]` returns a lazy `Column`, whose `__getitem__` contains (quoting 4.8.5, `arrow_dataset.py`):

```python
if self.source._format_type == "custom":
    # the formatting transform may require all columns
    source = self.source
else:
    source = self.source._fast_select_column(self.column_name)
return source[key][self.column_name]
```

LeRobot datasets always have `_format_type == "custom"` (from the `set_transform` call above), so the column-first "fast" branch literally routes to `hf_dataset[indices][key]`, the exact same code path as the row-first fallback. Both fetch and decode **all** columns of the queried rows. On image-type datasets (images embedded in parquet, e.g. v3.0-converted LIBERO) that means decoding `chunk_size × n_cameras` PNGs per sample just to read a `(50, 7)` float tensor.

In `datasets` 4.3.0 this guard did not exist: `Column.__getitem__` always went through `_fast_select_column`, a zero-copy single-column projection. That is why the column-first switch in #2408 benchmarked well at the time. It was borrowing an internal fast path that `datasets` 4.4 later disabled for custom-format datasets (a reasonable correctness fix on their side, since a user transform may need all columns). The version-dependent numbers in #3523 bracket the same change.

## Why "update `datasets`" seemed right, and why it no longer holds

This bug has changed shape twice, which is why the upgrade cure looks proven but isn't:

| When | Event | Flavor |
|---|---|---|
| Nov 2025 | #2408 switches to column-first (fixing the row-first N+1 of #2282), benchmarked on `datasets` 4.0–4.3 | fast: the lazy-`Column` fast path did the projection internally |
| Nov 2025 | #2549: user on `datasets` **3.6.0**, where `ds[key]` loads the whole column into memory → 2998 ms/query; upgrading to ≥ 4.0 fixed it; issue correctly closed | **< 4.0 flavor**: whole-column load |
| `datasets` 4.4.0 | the custom-format guard above is added | **≥ 4.4 flavor**: full-row fetch + decode for any dataset with a transform, i.e. every LeRobot dataset |
| Feb 2026 | this issue (#2895) | ≥ 4.4 flavor |
| Mar 2026 | #3210 closed with "with modern `datasets`, the current code should be fine", extrapolating from the #2549 precedent | valid for the < 4.0 flavor, not for this one |
| May 2026 | #3523 benchmarks the 4.3 → 4.4 cliff; #3558 hoists row-batch fetches | see "Relation to #3558" below |

So upgrading `datasets`, as @lhoestq suggested above, genuinely cured the < 4.0 flavor. But it cannot cure the ≥ 4.4 flavor: the fast path is now disabled *by design* whenever a transform is set, and LeRobot always sets one. The benchmark below is on `datasets` 4.8.5.

## Benchmark

lerobot 0.6.0, datasets 4.8.5, pyarrow 25.0.0, Python 3.12. Image-type dataset `HuggingFaceVLA/libero` (273,465 rows, 2 embedded-PNG camera columns). One 50-index action-chunk query on the `action` key; median of 12 interleaved rounds; all four methods verified `torch.equal`-identical:

| Strategy | median / query |
|---|---|
| `hf_dataset[key][indices]` (current try-branch) | 165.0 ms |
| `hf_dataset[indices][key]` (current fallback) | 162.8 ms |
| `list(hf_dataset.select(indices)[key])` (proposal in this issue) | 185.4 ms |
| cached `select_columns(key)` view (proposed below) | **0.8 ms** |

Two notes on the `select()` proposal:

- The verbatim `torch.stack(hf_dataset.select(indices)[key])` raises `TypeError` on `datasets` 4.8.5: `select(indices)[key]` now returns a lazy `Column`, which `torch.stack` rejects. (It was written against older `datasets` semantics, another example of how sensitive this code path is to `datasets` internals.)
- Benchmarked via `list(...)`, it is no faster than the current code on image-type datasets: the row subset still carries the custom transform, so extracting a column still fetches and decodes full rows. The 6.4 ms baseline in the issue description is consistent with a low-dim dataset without embedded images; for those, `select()` is fine, but for image-type datasets it is not.

## Proposed fix

Do the projection explicitly at the LeRobot level. Here we *know* `hf_transform_to_torch` is column-wise safe, an assumption `datasets` can never make for an arbitrary user transform:

```python
# cache per-key single-column views (zero-copy schema projection, built once)
view = self._column_views.setdefault(key, self.hf_dataset.select_columns(key))
result[key] = torch.stack(view[relative_indices][key])
```

This is robust to `datasets` version drift: it relies on schema projection plus index-based row gather, both stable public semantics, rather than on which internal path `ds[key][idx]` happens to take in a given release.

Production validation on our stack: per-sample `__getitem__` with a 50-step action chunk went from 200.7 ms to 4.4 ms (46×); in a 2×A800 DDP SmolVLA fine-tuning run, dataloader wait dropped from 38% of wall-clock to ~1% (`data_s` 0.16 → 0.008). Outputs verified bit-identical (`torch.equal`) against the old path over 324 samples including 24 episode-boundary cases.

Relation to #3558: hoisting a deduplicated row-batch fetch helps multi-key queries, but the shared batch is still a full-row fetch, so on image-type datasets it still decodes every embedded PNG. The two approaches are orthogonal and compose; per-key column views address the dominant cost.

Happy to open a PR with the fix plus tests (equivalence, and a contract test asserting the transform only ever receives the requested column, in the style of `test_dataset_reader.py`) if that helps.

**Comment by lhoestq** [comment]
happy to improve the logic behing `set_transform` / `set_format` in `datasets` if needed btw

e.g. it could allow column-based transform/format and also chaining e.g. "torch" formatting and then user-defined transform

**Comment by Rick0525** [comment]
@lhoestq That would be great. A column-based declaration sounds like exactly the right long-term fix: if a transform can be declared column-safe at `set_transform` time, the guard in `Column.__getitem__` can keep the `_fast_select_column` path enabled for it, and every downstream library with a column-wise transform gets fast column access back automatically. The chaining idea also fits LeRobot's case well, since `hf_transform_to_torch` is essentially torch formatting plus small per-column conversions.

I think the two fixes are complementary rather than either/or, mainly for timing reasons:

- A `datasets`-side API would land in a future release. For reference, the guard is unchanged as of `datasets` 5.0.0, and `lerobot` currently pins `datasets>=4.7,<5`, so benefiting from it would need a new `datasets` release plus a version bump and a code change in LeRobot in any case.
- The LeRobot-side cached `select_columns` view works today across `datasets` 4.4 through 5.0, and it has a clear retirement path: once the official API exists, the cached views can be replaced by a one-line declaration at `set_transform` time. So it is a stopgap that is easy to delete later, not a workaround that stays forever.

@imstevenpmwork with the above in mind, would you be open to a PR with the cached-view fix plus the tests mentioned earlier (output equivalence against the current path, and a contract test asserting the transform only receives the requested column)? Happy to time it however works best for you, e.g. rebasing on top of #3558 if that lands first.
